### ISSUE TITLE
Performance comparisons?

### ISSUE THREAD

**Issue by FezVrasta** [planttheidea/crio#5]
May you provide some benchmark between Crio and the other alternatives?

Thanks.

**Comment by planttheidea** [comment]
Definitely on the roadmap, but I want to do my own performance tuning first. I spit this out in a 12 hour coding frenzy, and while it works as expected, there are a number of performance shortfalls that need to be shored up. I can guarantee the performance at this point will not match something like Immutable.

**Comment by yamalight** [comment]
I'd forked [immutable-benchmarks](https://github.com/MrCheater/immutable-benchmarks) and added Crio to it for comparison. You can find the code and resulting charts here: https://github.com/yamalight/immutable-benchmarks

**Comment by planttheidea** [comment]
thanks @yamalight, appreciate it. Actually surprised at the comparables, considering the infancy of the project crio didn't do too badly, although the `set` and `setIn` numbers obviously show room for improvement.

I released a bunch of performance enhancements this morning (current version is 2.2.0), and I can focus on optimizing `set` and `setIn` based on these results.

**Comment by yamalight** [comment]
great! can't wait to see the results :)

**Comment by planttheidea** [comment]
So I just released another batch of performance enhancements ... by my rough before and after estimates it looks like I was able to drop `set` by about 25% and `setIn` by about 40%. Still slower than I would like them to be, but still a statistically significant improvement. Also, by eliminating the double-calculation of the hash for object changes I was doing before, all methods that update the values (`concat`, `fill`, `splice`, etc) should see performance improvements as well.

A side note, though ... my baseline of comparison (at least specific to `set` and `setIn`) is `seamless-immutable`, because it is the only other library of the four that actually _freezes_ the object (making it truly immutable). This is why both of us are slower in these updates, because once an object is immutable ... its unchangeable, really what we're both doing is cloning the object with the new value present, which can get expensive in ridiculously extreme scenarios. This is an intentional architectural decision, as I consider the true immutability of the object more valuable in a vast majority of real-world use-cases.

I'll keep plugging away with my tweaks to improve the performance (and I welcome any PRs!), but I wanted to clarify that to set expectations for potential users.

**Comment by planttheidea** [comment]
Another batch of performance improvements, this time I was able to improve `setIn` specifically by ~`11%. I don't have access to my primary machine right now (and won't for a few days, I'm in the process of moving), but once I do I'll work on incorporating the benchmark comparisons into this repo.

**Comment by danielepolencic** [comment]
I think the reason why Crio is slower in `set` and `setIn` is because mori, Immutable.js & co all use data sharing when a value is mutated. However crio recreates a full new object every time.

**Comment by planttheidea** [comment]
@danielepolencic it is true, both `seamless-immutable` and `crio` place a heavier emphasis on JavaScript interoperability (ability to use like a standard object) and true immutability (through `Object.freeze`) than on update performance. It is also the reason why we both perform so much better with `get` and `getIn`, because for us it is basically a native get whereas for `immutable` and `mori` it is an operation. A good article that covers this is [found here](http://jlongster.com/Using-Immutable-Data-Structures-in-JavaScript).

That said, `crio` does its best to reuse objects whenever possible to minimize memory consumption and improve performance.

``` javascript
const foo = crio([{foo: 'bar'}]);
const fooBar = foo.push({bar: 'baz'});

console.log(foo === fooBar); // false, new array created with {bar: 'baz'} added
console.log(foo[0] === fooBar[0]); // true, as {foo: 'bar'} was unchanged
```

In the above example, `{foo: 'bar}` is untouched and reused when creating `fooBar` from `foo`, no additional object cloning or creation needed other than for the new value.

The same is true for any item that is unchanged as a result of an operation:

``` javascript
const mappedFoo = foo.map((item) => {
  return item.set('foo', 'bar');
});

console.log(mappedFoo === foo); // false, it is a new object
console.log(mappedFoo[0] === foo[0]); // true, the object was reused because its value did not change
```

Bottom line, though, is that there are trade-offs with using any of these libraries. `immutable` and `mori` have opaque structures and are not truly immutable, but offer better performance in update operations. `seamless-immutable` and `crio` have structures that reflect the natural data structures and are truly immutable, but have a performance degradation that becomes noticeable with large-scale update operations.

**Comment by planttheidea** [comment]
Alright just a quick update on this...

Still in the process of unpacking, but I've been able to make some headway in the speed department. Profiling reveals that the most time-consuming portion of the process is the hash creation, and I've been able to optimize that in a number of ways. I just ran some benchmarks on my POS laptop I've been developing on, and the discrepancy between `seamless-immutable` and `crio` has narrowed substantially. Obviously I will keep on plugging, but crio is pretty competitive, and the difference is basically unnoticeable in any real-world transaction volume.

As mentioned, the hashing part is the most time-consuming piece, but I deep it valuable to optimize object-sharing and checking for value equality between `crio` objects. If a far more brilliant mind than my own wants to submit a PR related to hashing an object based on values ... I'd be most appreciative.

That's all, just keeping everyone abreast. Happy `crio`ing!

**Comment by yamalight** [comment]
I've updated [benchmark](https://github.com/yamalight/immutable-benchmarks) to use lates crio version - looks much closer to seamless-immutable now! 👍

**Comment by planttheidea** [comment]
This has been open for a while, and per the most recent benchmarks I'd say performance is getting much better, so I'm going to close it. Obviously performance improvements will continue, but no reason to leave this open ad infinitum. :)
